libraries/base/System/Mem/Weak.hs

```diff
@@ -140,3 +144,25 @@ A heap object is /reachable/ if:
  * It is the value or finalizer of a weak pointer object whose key is reachable.
 -}
 
+{- $notes
+
+A finalizers is not always called after its weak pointer\'s object becomes
+unreachable. There are two situations that can cause this:
+
+ * If the object becomes unreachable right before the program exits,
+   then GC may not be performed. Finalizers run during GC, so finalizers
```

Comment:
Almost correct, in fact what happens is the GC runs and creates a thread for the finalizer, but the program exits before the thread runs.  (could reference the other docs about not waiting for threads to finish before program exit).